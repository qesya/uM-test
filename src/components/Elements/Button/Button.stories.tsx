import { ArrowCircleLeftIcon, ArrowCircleRightIcon } from '@heroicons/react/outline';
import { Meta, Story } from '@storybook/react';
import React from 'react';

import { Button, ButtonProps } from './Button';

const meta: Meta = {
	title: 'Components/Elements/Button',
	component: Button,
	parameters: {
		control: {
			expanded: true,
		},
	},
};

export default meta;

const startIcon = <ArrowCircleLeftIcon className='h-6 w-6 mr-2' />;
const endIcon = <ArrowCircleRightIcon className='h-6 w-6 ml-2' />;

const Template: Story<ButtonProps> = (props) => <Button {...props} />;

export const Primary = Template.bind({});
Primary.args = {
	children: 'Primary Button',
	variant: 'primary',
};

export const Secondary = Template.bind({});
Secondary.args = {
	children: 'Secondary Button',
	variant: 'secondary',
};

export const Tertiary = Template.bind({});
Tertiary.args = {
	children: 'Tertiary Button',
	variant: 'tertiary',
};

export const PrimaryStartIcon = Template.bind({});
PrimaryStartIcon.args = {
	children: 'Primary Button',
	variant: 'primary',
	startIcon: startIcon,
};

export const PrimaryEndIcon = Template.bind({});
PrimaryEndIcon.args = {
	children: 'Primary Button',
	variant: 'primary',
	endIcon: endIcon,
};

export const Disabled = Template.bind({});
Disabled.args = {
	children: 'Primary Button',
	variant: 'primary',
	disabled: true,
};
