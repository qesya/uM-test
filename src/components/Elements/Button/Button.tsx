import clsx from 'clsx';
import { ButtonHTMLAttributes, forwardRef } from 'react';

import { Spinner } from '@/components/Elements/Spinner';

const variants = {
	primary:
		'bg-near-black text-near-white hover:bg-near-black-300 hover:text-near-white active:bg-blue-700 hover:text-near-white focus:outline-none focus:ring focus:ring-blue-200',
	secondary:
		'bg-orange-500 text-near-black hover:bg-orange-300 hover:text-near-black active:bg-orange-700 active:text-near-black focus:outline-none focus:ring focus:ring-orange-200',
	tertiary:
		'bg-near-black-100 text-near-black shadow-[inset_0px_0px_0px_2px] shadow-near-black-500',
	disabled: 'bg-grey-100 text-grey-300',
	textOnly: 'border-0 font-normal',
};

const sizes = {
	xs: 'px-2.5 py-1.5 text-xs rounded',
	sm: 'px-3 py-2 text-sm rounded-md',
	md: 'px-8 py-2 text-sm rounded-md',
	lg: 'px-8 py-2 text-base rounded-md',
	xlg: 'px-6 py-3 text-base rounded-md',
};

type IconProps =
	| { startIcon: React.ReactElement; endIcon?: never }
	| { endIcon: React.ReactElement; startIcon?: never }
	| { endIcon?: undefined; startIcon?: undefined };

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
	variant?: keyof typeof variants;
	size?: keyof typeof sizes;
	isLoading?: boolean;
} & IconProps;

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
	(
		{
			type = 'button',
			className = '',
			variant = 'primary',
			size = 'md',
			name = '',
			isLoading = false,
			disabled = false,
			startIcon,
			endIcon,
			...props
		},
		ref
	) => {
		return (
			<button
				ref={ref}
				data-testid='btn-test'
				type={type}
				className={clsx(
					'inline-flex items-center font-bold uppercase',
					disabled ? variants['disabled'] : variants[variant],
					sizes[size],
					className
				)}
				name={name}
				disabled={disabled}
				{...props}
			>
				{isLoading && <Spinner size='sm' className='text-current' />}
				{!isLoading && startIcon}
				<span className={clsx('mt-[1px]', isLoading ? 'ml-2' : '')}>{props.children}</span>
				{!isLoading && endIcon}
			</button>
		);
	}
);

Button.displayName = 'Button';
