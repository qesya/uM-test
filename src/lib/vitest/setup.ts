import '@testing-library/jest-dom';
//Fetch polyfill for RTK Query
import 'whatwg-fetch';
