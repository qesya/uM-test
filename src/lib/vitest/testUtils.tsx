/* eslint-disable import/export */
import { cleanup, render } from '@testing-library/react';
import React from 'react';
import { afterEach } from 'vitest';

import { AppProvider } from '@/providers/app';

afterEach(() => {
	cleanup();
});

const customRender = (ui: React.ReactElement, options = {}) => {
	render(ui, {
		wrapper: AppProvider,
		...options,
	});
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
// Override render export from @testing-library/react
export { customRender as render };
