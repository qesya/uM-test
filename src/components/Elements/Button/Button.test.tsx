/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */
import { fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';

import { Button } from './Button';

describe('Button', () => {
	it('should render the Button', async () => {
		const mockFn = vi.fn(() => {});
		const testBtn = 'test';

		const result = render(<Button onClick={mockFn}>{testBtn}</Button>);
		expect(result).toMatchSnapshot();
	});

	it('should press the onClick', async () => {
		const mockFn = vi.fn(() => {});
		const testBtn = 'test';
		const { getByTestId } = render(<Button onClick={mockFn}>{testBtn}</Button>);
		fireEvent.click(getByTestId('btn-test'));
		expect(mockFn).toHaveBeenCalled();
	});
});
