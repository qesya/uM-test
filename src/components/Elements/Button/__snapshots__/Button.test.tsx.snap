// Vitest Snapshot v1

exports[`Button > should render the Button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="inline-flex items-center font-bold uppercase bg-near-black text-near-white hover:bg-near-black-300 hover:text-near-white active:bg-blue-700 hover:text-near-white focus:outline-none focus:ring focus:ring-blue-200 px-8 py-2 text-sm rounded-md"
        data-testid="btn-test"
        name=""
        type="button"
      >
        <span
          class="mt-[1px]"
        >
          test
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="inline-flex items-center font-bold uppercase bg-near-black text-near-white hover:bg-near-black-300 hover:text-near-white active:bg-blue-700 hover:text-near-white focus:outline-none focus:ring focus:ring-blue-200 px-8 py-2 text-sm rounded-md"
      data-testid="btn-test"
      name=""
      type="button"
    >
      <span
        class="mt-[1px]"
      >
        test
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
